README.md
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/1aa1a2fafa1a41e1bdcba6e453fcba1f)](https://app.codacy.com/app/DigitalCoin1/SperoCoin?utm_source=github.com&utm_medium=referral&utm_content=DigitalCoin1/SperoCoin&utm_campaign=Badge_Grade_Dashboard)
[![Build Status](https://travis-ci.org/DigitalCoin1/SperoCoin.svg?branch=master)](https://travis-ci.org/DigitalCoin1/SperoCoin)
<br>
[![N|Solid](https://i.imgur.com/0BOgWZy.png)](https://sperocoin.org/)

Version: 2.7.0.6
=====================================

SperoCoin is a cryptocurrency developed with the aim of creating a decentralized ecosystem whose main objective is the application of SOCIAL SUSTAINABILITY.

Social sustainability implies generating income without losing focus on reducing social inequalities, with the aim of improving society's quality of life.
However, this does not mean that the company will fail to profit or run after profit. Instead, the idea is to create a business model in which profit is no longer above everything else, but to make the company also start to worry about its impact on the society in which it operates.

In addition to this project seeking social sustainability, we also aim at the privacy of payments, using the TOR network to forward transactions.

The integration with TOR ensures that users are always on a secure environment. There is no client-side configuration needed. Our client will automatically join the TOR network and connect to available TOR nodes on the network (including our seed nodes). After a successful connection to the SperoCoin network, you may check your displayed IP using the SperoCoin client console, where you will see that your client is assigned a TOR hidden service address (.onion address).

#### Pluggable Transports

>SperoCoin has options for advanced users to configure PTs such as obfs4 & Meek. 
>
>Instructions:
>
>[Setup OBFS4]( https://github.com/DigitalCoin1/SperoCoin/blob/master/doc/setup-obfs4.md )
>
>[Setup Meek]( https://github.com/DigitalCoin1/SperoCoin/blob/master/doc/setup-meek.md )


## Specifications

| Specification | Value |
|:-----------|:-----------|
| Consensus Algorithm | `PoW & PoS` |
| Mining Algorithm | `X13` |
| Confirmations | `3 tx` |
| Block Confirm. | `5 minted confirms` |
| Port | `55680` |
| RPC Port | `55681` |


##### PoW:

> - 60 sec block interval.
> - Difficulty readjustment every block.


| Blocks Interval | Block Reward |
|:-----------|:-----------|
| 1 - 2 | `50000` |
| 3 - 33330 | `2` |
| 33331 - 1149999 | `0.05 + 0.015 Foundation` |
| 1150000 - 1339199 | `0.025` |
| 1339200 - 1339211 | `0.025 + 10000 Foundation` |
| 1339212 - 1625599 | `0.025` |
| 1625600 - 2151199 | `0.015` |
| 2151200 - INFINITE | `0.001` |


##### PoS:

> - 60 sec block interval.
> - Difficulty readjustment every block.
> - Minimum holding time before the PoS will be generated is 2 hours.
> - Maximum allowed accumulated coin is 12 hours.

| Blocks Interval | Block Reward |
|:-----------|:-----------|
| 1 - 263249 | `5% per year` |
| 263250 - 1149999 | `25% per year` |
| 1150000 - 1338999 | `0.25 per block + 0.25 Foundation` |
| 1150000 - 2151199 | `0.25 per block` |
| 2151200 - INFINITE | `0.25 per block` |


## Resources

:package: For software downloads, please see [Downloads](https://sperocoin.org/getStarted)

[Whitepaper](https://sperocoin.org/whitepaper/) 

[Roadmap](https://sperocoin.org/roadmap)

[Community](https://sperocoin.org/community)


## SperoCoin.conf file

Download [SperoCoin.conf](https://sperocoin.org/files/SperoCoin.conf.php?action=download) (right-click > save link as).

For more info on how setup your .conf file, please visit https://sperocoin.org/configurationFile


License :bookmark_tabs:
-------

SperoCoin is released under the terms of the MIT license. See [COPYING](COPYING) for more
information or see https://opensource.org/licenses/MIT.



Development Process
-------------------

The `master` branch is regularly built and tested, but is not guaranteed to be completely stable. 



changelog.md
CHANGELOG SPEROCOIN

>$0001 2017/09/26<br>
Initial Upload to GitHub<br>

>$0006 2017/10/01<br>
In this update, the SperoCoin-qt.pro file has been changed for compilation in the Linux environment.<br>

>$0007 2017/10/06<br>
Update README.md and add files: src/leveldb/Makefile<br>

>$0008 2017/10/08<br>
Update infos about end phase of PoW and start phase PoS;<br>
Update font of qt wallet;<br>
Update version;<br>
Update Protocol Others general's updates<br>

>$0010 2017/10/10<br>
Update for new implementations:<br>
→ Added option of cleaning of orphans blocks;<br>
→ Added display option "STAKE" in the "type" column inside the "Transactions" tab;<br>
→ Added "setban" node "" add | remove "(bantime)" / "listbanned" and "clearbanned" in the RPC console;<br>

>$0012 2017/10/11<br>
Ban Forked Peers who are stuck in a Getblocks Loop.<br>

>$0013 2017/10/22<br>
→ Update of strings names;<br>
→ Updating information in README.md<br>
→ Update Stake Info in QT  Wallet<br>

>$0015 2017/11/18<br>
→ Release of testnet mining;<br>
→ File Fix:<br>
   • main.cpp<br>
   • alter.cpp<br>
   • checkpoints.cpp<br>
   • checkpoints.h<br>

>$0016 2017/11/30<br>
Add node in the source:<br>
addnode=http://dnssperocoin.ddnsking.com:55680<br>

>$0017 2018/03/28<br>
→ Add RPC Command "GETWALLETINFO"<br>
→ Update of transaction maturity icons<br>
→ splash screen main update<br>
→ Updating the logo inside the Wallet<br>
→ Added hashGenesisBlockTestNet = 0x00001fe23e7d6951a2a07ad684ff208fab9110fad607fabd81482a1692382f8f in main.h file<br>
→ Added testnet block.nNonce in main.cpp file (block.nNonce = 15547;)<br>
→ Added static MapCheckpoints mapCheckpointsTestnet = (0, uint256("0x000006f4925ed7b889f847f24621390d4943466f091c3254ca1bd8becc517f7b")) in the checkpoints.cpp file<br>
→ Change the initial letter of the testnet addresses in the base58.h file (PUBKEY_ADDRESS_TEST = 125, // Initial address letter (63 = s) -TESTNET))<br>
→ Change the example initial letter in the file qt / sendcoinsdialog.cpp<br>
(ui-> lineEditCoinControlChange-> setPlaceholderText (tr ("Enter the SperoCoin address (e.g. Sjz75uKHzUQJnSdzvpiigEGxseKkDhQToX)"));)<br>
→ Changed version to 2.6.4.3<br>

>$0019 2018/07/15<br>
→ Added RPC command "getnetworkinfo" - Change in files: bitcoinrpc.cpp; bitcoinrpc.h; rpcblockchain.cpp; rpcnet.cpp<br>
→ Added RPC command "getblockchaininfo" - Change in files:bitcoinrpc.cpp; bitcoinrpc.h; rpcblockchain.cpp; rpcnet.cpp<br>
→ Added Wallet Status Information (Locked or Unlocked) in the RPC command "getinfo"<br>
→ Added new points in the network map updating the mapping up to block 130000- Change in file: checkpoints.cpp<br>
→ Changing copyrights, updating the start date of SperoCoin works, being: "Copyright © 2017 The SperoCoin developers" - Change in ".ts" files in the "src \ qt \ locale" folder<br>
→ Update of QT splashscreen images<br>
→ Changing QT fonts for "Century Gothic"<br>
→ Clear Orphans with automatic transaction page refresh, no need to restart wallet<br>
→ Changed version to 2.6.4.4<br>

>$0020 2018/07/31<br>
→ Add/Remove Nodes<br>

>$0021 2018/10/27<br>
→ Added new points in the network map updating the mapping up to block 165962- Change in file: checkpoints.cpp<br>
→ Performance Enhancements - Increased Speed ​​in Synchronization - Changes in files: main.cpp; main.h;<br>
→ Added new icons<br>
→ Change style of block download progress bar from "QWindowsXPStyle" to "QWindowsVistaStyle" - Changes in file: bitcoingui.cpp<br>
→ Changing the size of the icons in "STATUSBAR_ICONSIZE" from 16 to 64 - Changes in file: guiconstants.h<br>
→ Adding SperoCoin.conf file opening from the Console Window menu<br>
→ Changed description of mining type PoW<br>
→ Changed description of mining type PoS<br>
→ New PoS Mining Icon<br>
→ Changed version to 2.6.4.5<br>

>$0022 2018/12/08<br>
→ Alterada Versão para 2.6.4.6;<br>
→ Alterada versão do protocolo para mínimo 700001;<br>
→ Adicionada Introdução;<br>
→ Adicionado pontos de checagem de blocos até o bloco 180000<br>

>$0028 2019/05/31<br>
→ Alterada Versão para 2.6.4.7;<br>
→ Alterada versão do protocolo mínimo para 700002;<br>
→ Adicionada página Charity(Doações POS);<br>
→ Adicionados endereços Donate Foudation;<br>
→ Correção das strings que determinam as fases de mineração;<br>
→ Preparação para o ambiente de mineração PoW após o bloco 263250;<br>
→ Mudança dos ícones e logos;<br>
→ Adição do parâmetro -zapwallettxes;<br>

>$0029 2020/10/22<br>
→ [UPDATE] Nova versão 2.6.5.0;<br>
→ [UPDATE] Adição de lista de endereços de participantes do sistema de doações(CHARITY);<br>
→ [OPMIZATION] Limpeza e Remoção de arquivos desnecessários da base do código;<br>
→ [FEATURE] Compatibilidade com Openssl versões 1.0.2 e acima;<br>
→ [FEATURE] Compatibilidade com Boost 1.66 e 1.70;<br>
→ [FEATURE] Implementação de auto-completar na janela de depuração;<br>
→ [FIX] getbalance() bug;<br>
→ [FIX] TransactionView: Cálculo do último mês;<br>
→ [UPDATE] Array de seed nodes;<br>
→ [UPDATE] Checkpoints;<br>
→ [UPDATE] Arquivos de Compilação SperoCoin-qt.pro e makefile.mingw usados em compilação Windows<br>

>$0030 2021/01/29<br>
→ [UPDATE] Nova versão 2.6.5.1;<br>
→ [UPDATE] Alterada versão do protocolo para mínimo 700003;<br>
→ [GUI] Novo Design QT;<br>
→ [GUI] Nova Splash Screen;<br>
→ [FIX] Dev fee;<br>
→ [FIX] Staking Dust;<br>
→ [FIX] Compilações MinGW/QT;<br>
→ [UPDATE] Novas premiações PoW e PoS;<br>
→ [UPDATE] Novos endereços Foundation PoS;<br>
→ [NEW] Halving Blocks<br>

>$0031 2021/03/18<br>
→ [NEW] Checkpoints;<br>
→ [FIX] Mineração PoS;<br>
→ [BUMP] Alterada versão do protocolo mínimo para 700004;<br>
→ [BUMP] Nova versão 2.6.5.2;<br>

>$0032 2021/10/14<br>
→ [NEW] Checkpoints;<br>
→ [REMOVE] Pagamentos para a Foundation PoS;<br>
→ [NEW] Fase especial de mineração para a Foundation PoW;<br>
→ [FORK] Preparação do código para o hardfork;<br>

src/util.h
// Copyright (c) 2009-2010 Satoshi Nakamoto
// Copyright (c) 2009-2017 The Bitcoin Core developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

/**
 * Server/client environment: argument handling, config file parsing,
 * logging, thread wrappers, startup time
 */
#ifndef BITCOIN_UTIL_H
#define BITCOIN_UTIL_H

#if defined(HAVE_CONFIG_H)
#include <config/bitcoin-config.h>
#endif

#include <string.h>
#include <compat.h>
#include <fs.h>
#include <sync.h>
#include <tinyformat.h>
#include <utiltime.h>
#include <amount.h>

#ifndef WIN32
#include <sys/resource.h>
#endif

#include <atomic>
#include <exception>
#include <map>
#include <stdint.h>
#include <string>
#include <vector>

#include <zlib.h>
#include <boost/scoped_array.hpp>


/* These are all highly standard and portable headers. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* This is for mkdir(); this may need to be changed for some platforms. */
#include <sys/stat.h>  /* For mkdir() */

#include <boost/signals2/signal.hpp>
#include <boost/thread/condition_variable.hpp> // for boost::thread_interrupted

// Application startup time (used for uptime calculation)
int64_t GetStartupTime();

static const bool DEFAULT_LOGTIMEMICROS = false;
static const bool DEFAULT_LOGIPS        = false;
static const bool DEFAULT_LOGTIMESTAMPS = true;
extern const char * const DEFAULT_DEBUGLOGFILE;

/** Signals for translation. */
class CTranslationInterface
{
public:
    /** Translate a message to the native language of the user. */
    boost::signals2::signal<std::string (const char* psz)> Translate;
};

extern bool fPrintToConsole;
extern bool fPrintToDebugLog;

extern bool fLogTimestamps;
extern bool fLogTimeMicros;
extern bool fLogIPs;
extern std::atomic<bool> fReopenDebugLog;
extern CTranslationInterface translationInterface;

extern const char * const BITCOIN_CONF_FILENAME;
extern const char * const BITCOIN_PID_FILENAME;

extern std::atomic<uint32_t> logCategories;

/**
 * Translation function: Call Translate signal on UI interface, which returns a boost::optional result.
 * If no translation slot is registered, nothing is returned, and simply return the input.
 */
inline std::string _(const char* psz)
{
    boost::optional<std::string> rv = translationInterface.Translate(psz);
    return rv ? (*rv) : psz;
}

void SetupEnvironment();
bool SetupNetworking();

struct CLogCategoryActive
{
    std::string category;
    bool active;
};

namespace BCLog {
    enum LogFlags : uint32_t {
        NONE        = 0,
        NET         = (1 <<  0),
        TOR         = (1 <<  1),
        MEMPOOL     = (1 <<  2),
        HTTP        = (1 <<  3),
        BENCH       = (1 <<  4),
        ZMQ         = (1 <<  5),
        DB          = (1 <<  6),
        RPC         = (1 <<  7),
        ESTIMATEFEE = (1 <<  8),
        ADDRMAN     = (1 <<  9),
        SELECTCOINS = (1 << 10),
        REINDEX     = (1 << 11),
        CMPCTBLOCK  = (1 << 12),
        RAND        = (1 << 13),
        PRUNE       = (1 << 14),
        PROXY       = (1 << 15),
        MEMPOOLREJ  = (1 << 16),
        LIBEVENT    = (1 << 17),
        COINDB      = (1 << 18),
        QT          = (1 << 19),
        LEVELDB     = (1 << 20),
        POS         = (1 << 21),
        STAKE       = (1 << 22),
        STEALTH     = (1 << 23),
        WALLET      = (1 << 24),
        SPEROVAULT   = (1 << 25),
        SPEROSEND    = (1 << 26),
		CHECKPOINT  = (1 << 27),
        FASTSYNC   = (1 << 28),
        ALL         = ~(uint32_t)0,
    };
}
/** Return true if log accepts specified category */
static inline bool LogAcceptCategory(uint32_t category)
{
    return (logCategories.load(std::memory_order_relaxed) & category) != 0;
}

/** Returns a string with the log categories. */
std::string ListLogCategories();

/** Returns a vector of the active log categories. */
std::vector<CLogCategoryActive> ListActiveLogCategories();

/** Return true if str parses as a log category and set the flags in f */
bool GetLogCategory(uint32_t *f, const std::string *str);

/** Send a string to the log output */
int LogPrintStr(const std::string &str);

/** Get format string from VA_ARGS for error reporting */
template<typename... Args> std::string FormatStringFromLogArgs(const char *fmt, const Args&... args) { return fmt; }

static inline void MarkUsed() {}
template<typename T, typename... Args> static inline void MarkUsed(const T& t, const Args&... args)
{
    (void)t;
    MarkUsed(args...);
}

// Be conservative when using LogPrintf/error or other things which
// unconditionally log to debug.log! It should not be the case that an inbound
// peer can fill up a users disk with debug.log entries.

#ifdef USE_COVERAGE
#define LogPrintf(...) do { MarkUsed(__VA_ARGS__); } while(0)
#define LogPrint(category, ...) do { MarkUsed(__VA_ARGS__); } while(0)
#else
#define LogPrintf(...) do { \
    std::string _log_msg_; /* Unlikely name to avoid shadowing variables */ \
    try { \
        _log_msg_ = tfm::format(__VA_ARGS__); \
    } catch (tinyformat::format_error &fmterr) { \
        /* Original format string will have newline so don't add one here */ \
        _log_msg_ = "Error \"" + std::string(fmterr.what()) + "\" while formatting log message: " + FormatStringFromLogArgs(__VA_ARGS__); \
    } \
    LogPrintStr(_log_msg_); \
} while(0)

#define LogPrint(category, ...) do { \
    if (LogAcceptCategory((category))) { \
        LogPrintf(__VA_ARGS__); \
    } \
} while(0)
#endif

template<typename... Args>
bool error(const char* fmt, const Args&... args)
{
    LogPrintStr("ERROR: " + tfm::format(fmt, args...) + "\n");
    return false;
}

void PrintExceptionContinue(const std::exception *pex, const char* pszThread);
void FileCommit(FILE *file);
bool TruncateFile(FILE *file, unsigned int length);
int RaiseFileDescriptorLimit(int nMinFD);
void AllocateFileRange(FILE *file, unsigned int offset, unsigned int length);
bool RenameOver(fs::path src, fs::path dest);
bool LockDirectory(const fs::path& directory, const std::string lockfile_name, bool probe_only=false);

/** Release all directory locks. This is used for unit testing only, at runtime
 * the global destructor will take care of the locks.
 */
void ReleaseDirectoryLocks();

bool TryCreateDirectories(const fs::path& p);
fs::path GetDefaultDataDir();
const fs::path &GetDataDir(bool fNetSpecific = true);
void ClearDatadirCache();
fs::path GetConfigFile(const std::string& confPath);
#ifndef WIN32
fs::path GetPidFile();
void CreatePidFile(const fs::path &path, pid_t pid);
#endif
#ifdef WIN32
fs::path GetSpecialFolderPath(int nFolder, bool fCreate = true);
#endif
fs::path GetDebugLogPath();
bool OpenDebugLog();
void ShrinkDebugFile();
void runCommand(const std::string& strCommand);

inline bool IsSwitchChar(char c)
{
#ifdef WIN32
    return c == '-' || c == '/';
#else
    return c == '-';
#endif
}

class ArgsManager
{
protected:
    mutable CCriticalSection cs_args;
    std::map<std::string, std::string> mapArgs;
    std::map<std::string, std::vector<std::string>> mapMultiArgs;
public:
    void ParseParameters(int argc, const char*const argv[]);
    void ReadConfigFile(const std::string& confPath);

    /**
     * Return a vector of strings of the given argument
     *
     * @param strArg Argument to get (e.g. "-foo")
     * @return command-line arguments
     */
    std::vector<std::string> GetArgs(const std::string& strArg) const;

    /**
     * Return true if the given argument has been manually set
     *
     * @param strArg Argument to get (e.g. "-foo")
     * @return true if the argument has been set
     */
    bool IsArgSet(const std::string& strArg) const;

    /**
     * Return string argument or default value
     *
     * @param strArg Argument to get (e.g. "-foo")
     * @param strDefault (e.g. "1")
     * @return command-line argument or default value
     */
    std::string GetArg(const std::string& strArg, const std::string& strDefault) const;

    /**
     * Return integer argument or default value
     *
     * @param strArg Argument to get (e.g. "-foo")
     * @param nDefault (e.g. 1)
     * @return command-line argument (0 if invalid number) or default value
     */
    int64_t GetArg(const std::string& strArg, int64_t nDefault) const;

    /**
     * Return boolean argument or default value
     *
     * @param strArg Argument to get (e.g. "-foo")
     * @param fDefault (true or false)
     * @return command-line argument or default value
     */
    bool GetBoolArg(const std::string& strArg, bool fDefault) const;

    /**
     * Set an argument if it doesn't already have a value
     *
     * @param strArg Argument to set (e.g. "-foo")
     * @param strValue Value (e.g. "1")
     * @return true if argument gets set, false if it already had a value
     */
    bool SoftSetArg(const std::string& strArg, const std::string& strValue);

    /**
     * Set a boolean argument if it doesn't already have a value
     *
     * @param strArg Argument to set (e.g. "-foo")
     * @param fValue Value (e.g. false)
     * @return true if argument gets set, false if it already had a value
     */
    bool SoftSetBoolArg(const std::string& strArg, bool fValue);

    // Forces an arg setting. Called by SoftSetArg() if the arg hasn't already
    // been set. Also called directly in testing.
    void ForceSetArg(const std::string& strArg, const std::string& strValue);
};

extern ArgsManager gArgs;

/**
 * Format a string to be used as group of options in help messages
 *
 * @param message Group name (e.g. "RPC server options:")
 * @return the formatted string
 */
std::string HelpMessageGroup(const std::string& message);

/**
 * Format a string to be used as option description in help messages
 *
 * @param option Option message (e.g. "-rpcuser=<user>")
 * @param message Option description (e.g. "Username for JSON-RPC connections")
 * @return the formatted string
 */
std::string HelpMessageOpt(const std::string& option, const std::string& message);

/**
 * Return the number of physical cores available on the current system.
 * @note This does not count virtual cores, such as those provided by HyperThreading
 * when boost is newer than 1.56.
 */
int GetNumCores();

void RenameThread(const char* name);

/**
 * .. and a wrapper that just calls func once
 */
template <typename Callable> void TraceThread(const char* name,  Callable func)
{
    std::string s = strprintf("sperocoin-%s", name);
    RenameThread(s.c_str());
    try
    {
        LogPrintf("%s thread start\n", name);
        func();
        LogPrintf("%s thread exit\n", name);
    }
    catch (const boost::thread_interrupted&)
    {
        LogPrintf("%s thread interrupt\n", name);
        throw;
    }
    catch (const std::exception& e) {
        PrintExceptionContinue(&e, name);
        throw;
    }
    catch (...) {
        PrintExceptionContinue(nullptr, name);
        throw;
    }
}

std::string CopyrightHolders(const std::string& strPrefix);

//! Substitute for C++14 std::make_unique.
template <typename T, typename... Args>
std::unique_ptr<T> MakeUnique(Args&&... args)
{
    return std::unique_ptr<T>(new T(std::forward<Args>(args)...));
}

#ifdef WIN32
inline void SetThreadPriority(int nPriority)
{
    SetThreadPriority(GetCurrentThread(), nPriority);
}
#else

#define THREAD_PRIORITY_LOWEST          PRIO_MAX
#define THREAD_PRIORITY_BELOW_NORMAL    2
#define THREAD_PRIORITY_NORMAL          0
#define THREAD_PRIORITY_ABOVE_NORMAL    0

inline void SetThreadPriority(int nPriority)
{
    // It's unclear if it's even possible to change thread priorities on Linux,
    // but we really and truly need it for the generation threads.
#ifdef PRIO_THREAD
    setpriority(PRIO_THREAD, 0, nPriority);
#else
    setpriority(PRIO_PROCESS, 0, nPriority);
#endif
}
#endif

inline uint32_t ByteReverse(uint32_t value)
{
    value = ((value & 0xFF00FF00) >> 8) | ((value & 0x00FF00FF) << 8);
    return (value<<16) | (value>>16);
}

const std::string currentDateTime();
#endif // BITCOIN_UTIL_H
